[k8s] Support for switching k8s contexts

romilbhardwaj opened the pull request:
Previously, if the user switched kube contexts and tried `exec`/`down` or other operations on a cluster, they would fail with a ClusterIdentityMismatch error. 

This PR changes the behavior to allow the user to operate on clusters even if the `current-context` in the kubeconfig is different from the context that was used to launch the cluster.

This is implemented by storing kubecontext information as a part of the cluster config and making all our Kubernetes API calls context-sensitive. Since this effectively requires switching "identities", we also add a `cloud.get_supported_identities` method to get a list of all identities SkyPilot can switch to. 

I considered using a less invasive approach (which does not require changing all API calls to be context sensitive), one which used a python contextmanager to temporarily set the context globally in `adaptors.kubernetes._load_config()`, but that was quickly discarded since many SkyPilot operations (e.g., down) are run in parallel and thread safety becomes hard.

TODO:
- [x] Extensive testing, incl. all kubernetes smoke tests and backward compatibility

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Manual tests with 
- [x] New smoke test - `test_kubernetes_context_switch` 
- [x] Other smoke tests and backward compatibility

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -1559,6 +1559,7 @@ def check_owner_identity(cluster_name: str) -> None:
 
     cloud = handle.launched_resources.cloud
     current_user_identity = cloud.get_current_user_identity()
+    supported_user_identities = cloud.get_supported_identities()
```
Should we directly make `get_current_user_identity` as a list, instead of adding a new API?

romilbhardwaj replied:
I was a little confused by our implementation (and perhaps there's a bug there) which made me add this new method.

Currently, we iterate on the `zip(owner_identity, current_user_identity)` to check identities:

https://github.com/skypilot-org/skypilot/blob/2e545b8db70991bc74d112d5070fa908078f9cdb/sky/backends/backend_utils.py#L1578-L1586

Say the user had launched the cluster with just one identity configured in their kubeconfig, and thus `owner_identity` is `["myid1"]`. 

Say they updated their kubeconfig and now they have 3 identities in `current_user_identity`, the original one being 5th in the list: `["myid3", "myid2", "myid1"]`. 

`zip(owner_identity, current_user_identity)` will now be `["myid1", "myid3"]`, which will fail the check we have.

```
>>> x=["myid1"]
>>> y=["myid3", "myid2", "myid1"]
>>> list(zip(x,y))
[('myid1', 'myid3')]
```

For k8s this logic should be changed to [itertools.product](http://docs.python.org/library/itertools.html#itertools.product) instead of `zip`, but I think that will affect AWS implementation, which has a stricter ordering requirement.

Michaelvll replied:
The reason we have a list for the `current_user_identity` and `owner_identity` is that it can be fine for a cloud owner with a different cloud account but from the same project to operate a VM, for example, we allows the following to match
```python
owner_identity = ['myaccount1', 'myproject']
current_identity = ['myaccount2', 'myproject']
for i, (owner, 
         current) in enumerate(zip(owner_identity, 
                                   current_user_identity)): 

```

In order to support multiple `current_identity`, we can do something like the following:
```python
owner_identity = ['myaccount1', 'myproject']
current_identity = [['myaccount2', 'myproject']]
for available_identity in current_identities:
  for i, (owner, 
         current) in enumerate(zip(owner_identity, 
                                                      available_identity)): 

```

And for k8s, it can be:
```python
owner_identity = ['mycontext']
current_identity = [['mycontext1'], ['mycontext2']]
for available_identity in current_identities:
  for i, (owner, 
         current) in enumerate(zip(owner_identity, 
                                                      available_identity)): 

```

Michaelvll reviewed (commented):
Thanks for adding this @romilbhardwaj! Looks mostly good to me with minor comments for an interface.

romilbhardwaj commented:
Thanks @Michaelvll! I've updated our identity logic:

* Renamed `get_current_user_identity` to `get_active_user_identity`
* Added `get_user_identities` to fetch all identities. 

We can roll these into one method as well (i.e., just have `get_user_identities` and replace usage of  `cloud.get_active_user_identity()` with `cloud.get_user_identities()[0]`), but I figured this is a little cleaner.

romilbhardwaj commented:
Tested:
- [x] `sky launch` on `context1`, switch to `context2`, launch on `context2`, run `sky exec/queue/queue/logs/cancel` on cluster running in `context1`
- [x] `sky launch` on `context1`, switch to `context2`, launch on `context2`, delete `context1` from kubeconfig and make sure IdentityError is thrown
- [ ] `sky jobs launch` on `context1`, switch to `context2`, launch on `context2`, run `sky jobs launch` on controller running in `context1`
- [ ] `sky serve up` on `context1`, switch to `context2`, launch on `context2`, run `sky serve up` on controller running in `context1`
- [ ] New smoke test `test_kubernetes_context_switch`
- [ ] All smoke tests

romilbhardwaj commented:
This PR now also [fixes](https://github.com/skypilot-org/skypilot/pull/3913/commits/191e998d974bf9f6cc98080956ccebe0ac7d63d4) a bug introduced by #3897 where if a namespace was passed but no context was set (e.g., when happen inside a controller pod when using incluster auth) to `kubernetes-port-forward-proxy-command.sh`, the namespace arg would be incorrectly assumed to be the context arg and ssh would fail.

romilbhardwaj commented:
Backward compatibility is now fixed. 

TODO:
- [ ] `sky launch` from a different context leaks pods in the original context since the cluster owner changes. We may need to add logic to disallow owner change when cluster already exists in state.db.

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -1558,58 +1558,65 @@ def check_owner_identity(cluster_name: str) -> None:
+                    return  # The user identity matches.
+        # Generate error message if no match found
+        if len(user_identities) == 1:
+            err_msg = f'the activated account is {user_identities[0]!r}.'
```
```suggestion
            err_msg = f'the activated identity is {user_identities[0]!r}.'
```

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -1558,58 +1558,65 @@ def check_owner_identity(cluster_name: str) -> None:
+        if len(user_identities) == 1:
+            err_msg = f'the activated account is {user_identities[0]!r}.'
+        else:
+            err_msg = (f'available accounts are {user_identities!r}.')
```
```suggestion
            err_msg = (f'available identities are {user_identities!r}.')
```

Michaelvll commented on sky/clouds/cloud.py:
```diff
@@ -489,13 +490,41 @@ def get_current_user_identity(cls) -> Optional[List[str]]:
             return None
         return ', '.join(user_identity)
 
+    @classmethod
+    def get_user_identities(cls) -> Optional[List[List[str]]]:
+        """Returns all available user identities of this cloud.
+
+        See get_active_user_identity for definition of user identity.
+
+        This method returns a list of user identities, with the current active
+        identity being the first element. Most clouds have only one identity
+        available, so the returned list will only have one element: the current
+        active identity.
+
+        However, some clouds (e.g., Kubernetes) can have multiple current
+        identities (e.g., multiple contexts configured in kubeconfig) that
+        can be dynamically switched, so the list can have multiple elements.
+
+        Example return values:
+            - AWS: [[UserId, AccountId]]
+            - GCP: [[email address + project ID]]
+            - Azure: [[email address + subscription ID]]
+            - Kubernetes: [[current active context], [context 2], ...]
+
+        Returns:
+            None if the cloud does not have a concept of user identity;
+            otherwise all the user identities.
+        """
+        active_identity = cls.get_active_user_identity()
+        return [active_identity] if active_identity is not None else None
```
I found calling `get_active_user_identity` within `get_user_identities` a bit weird to me. Should we instead make `get_user_identities` always return a list of identities with the active one being the first one, and having `get_active_user_identity` to index the returned list of identities? 

With that, we can also add `TODO` in AWS and GCP's implementation: `Return a list of identities in the profile, when we support automatically switching context.`

romilbhardwaj replied:
I see what you mean, fixed now!

Michaelvll commented on sky/provision/kubernetes/utils.py:
```diff
@@ -58,7 +58,10 @@
+# We add a version suffix to the port-forward proxy command to ensure backward
+# compatibility and avoid overwriting the older version.
+PORT_FORWARD_PROXY_CMD_PATH = ('~/.sky/'
+                               'kubernetes-port-forward-proxy-command-v2.sh')
```
nit: we can save the version of the proxy command in constants, if we expect it can change in the future

romilbhardwaj replied:
Ahh yeah we don't have a constants for kubernetes yet, but I've moved it to it's own variable for now

Michaelvll commented on tests/test_smoke.py:
```diff
@@ -1171,6 +1171,58 @@ def test_kubernetes_storage_mounts():
+        f'sky exec {name} "echo Success: sky exec works after context switch"',
+
+        # Test sky queue
+        f'sky queue {name}',
```
Should we add a test for SSH as well?

Michaelvll reviewed (commented):
Thanks for the fix @romilbhardwaj! Mostly looks good to me.

romilbhardwaj commented on sky/adaptors/kubernetes.py:
```diff
@@ -103,73 +96,53 @@ def _load_config():
     _configured = True
 
 
+@functools.lru_cache()
 @_api_logging_decorator('urllib3', logging.ERROR)
```
good point, fixed!

romilbhardwaj commented on sky/adaptors/kubernetes.py:
```diff
@@ -66,7 +59,7 @@ def wrapped(*args, **kwargs):
-def _load_config():
+def _load_config(context: Optional[str] = None):
     global _configured
     if _configured:
         return
```
ahh yes we can remove this since we now use lru_cache

romilbhardwaj commented on sky/authentication.py:
```diff
@@ -415,13 +419,14 @@ def setup_kubernetes_authentication(config: Dict[str, Any]) -> Dict[str, Any]:
         # Setup service for SSH jump pod. We create the SSH jump service here
         # because we need to know the service IP address and port to set the
         # ssh_proxy_command in the autoscaler config.
-        kubernetes_utils.setup_ssh_jump_svc(ssh_jump_name, namespace,
+        kubernetes_utils.setup_ssh_jump_svc(ssh_jump_name, namespace, context,
                                             service_type)
```
It is still used when `kubernetes.networking` is `nodeport` (we will deprecate that soon)

Michaelvll commented on sky/clouds/lambda_cloud.py:
```diff
@@ -258,8 +258,8 @@ def get_credential_file_mounts(self) -> Dict[str, str]:
-    def get_current_user_identity(cls) -> Optional[List[str]]:
-        # TODO(ewzeng): Implement get_current_user_identity for Lambda
+    def get_user_identities(cls) -> Optional[List[List[str]]]:
+        # TODO(ewzeng): Implement get_active_user_identity for Lambda
```
```suggestion
        # TODO(ewzeng): Implement get_user_identities for Lambda
```

Michaelvll commented on sky/clouds/scp.py:
```diff
@@ -337,8 +337,8 @@ def get_credential_file_mounts(self) -> Dict[str, str]:
-    def get_current_user_identity(cls) -> Optional[List[str]]:
-        # TODO(jgoo1): Implement get_current_user_identity for SCP
+    def get_user_identities(cls) -> Optional[List[List[str]]]:
+        # TODO(jgoo1): Implement get_active_user_identity for SCP
```
```suggestion
        # TODO(jgoo1): Implement get_user_identities for SCP
```

Michaelvll commented on sky/utils/kubernetes/rsync_helper.sh:
```diff
@@ -1,7 +1,16 @@
-# When using pod@namespace, rsync passes args as: {us} -l pod namespace 
+# When using pod@namespace/context, rsync passes args as: {us} -l pod namespace+context
```
Seems we are using `namespace+context` when we are calling rsync. Should we update this comment?
```suggestion
# When using pod@namespace+context, rsync passes args as: {us} -l pod namespace+context
```

Michaelvll reviewed (approved):
Thanks @romilbhardwaj for the great effort! LGTM.

romilbhardwaj commented:
Verified it works with the job controller as well and `sky launch` works across contexts. Kubernetes smoke tests pass, merging now.
